platforms/jvm/plugins-jvm-test-suite/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java

```diff
@@ -305,12 +191,19 @@ public void useJUnit(String version) {
 
     @Override
     public void useJUnit(Provider<String> version) {
-        setFrameworkTo(TestingFramework.JUNIT4, version);
+        useJUnit(parameters -> parameters.getVersion().set(version));
+    }
+
+    private void useJUnit(Action<JUnit4ToolchainParameters> action) {
+        setToolchainAndConfigure(JUnit4TestToolchain.class, parameters -> {
+            parameters.getVersion().convention(JUnit4TestToolchain.DEFAULT_VERSION);
```

Comment:
Thoughts on moving this line up into `useJUnit()` so that for the non-version-accepting methods we are not needlessly setting the convention just to override it immediately with a `set`?

Is there value in setting a convention even if we override it immediately with a `set`?

Reply:
So, for this PR, we could do that because the only actions passed here are version-setting actions.  However, if we make this method public in order to make other things user-configurable on the parameters object (for instance, the `TestFrameworkOptions` method or `TestEngine` instances) we would need this here, to ensure that the convention was set since we would no longer be able to assume the provided action is setting a version.  I agree that this is _slightly_ inefficient but I don't think there's any obvious disadvantage to doing it.  I'd rather just leave it than move it and then potentially move it back later.